src/python/lib/ensembl/compara/filecmp/dircmp.py

```diff
@@ -0,0 +1,216 @@
+
+        Note:
+            Files of each folder will be under the key ``.``.
+
+        Args:
+            path: Root path.
+
+        """
+        if os.path.isfile(path):
+            tree_dict = {'.': set(os.path.basename(path))}  # type: Dict
+        else:
+            tree_dict = {'': {}}
+            # Breadth First traversal of the directory tree
+            for dirpath, dirnames, filenames in os.walk(path):
+                # Transform dirpath into a relative path (from the root path)
+                dirpath = dirpath.replace(path, '')
+                subtree = reduce(operator.getitem, dirpath.split(os.sep), tree_dict)
+                # Initialise the subdirectories to guarantee reduce() will find them
+                subtree.update({name: {} for name in dirnames})
+                subtree['.'] = set(filenames)
+            tree_dict = tree_dict['']
+        return tree_dict
+
+    @staticmethod
+    def _prune_tree(root: Dict, paths: Union[str, List] = None, raise_err: bool = False) -> Dict:
+        """Returns a pruned directory tree that only includes the given paths.
+
+        Note:
+            If `paths` is empty, it returns `root`.
+
+        Args:
+            root: Directory tree.
+            paths: Relative directory path(s) to include.
+            raise_err: Raise exception flag.
+
+        Raises:
+            ValueError: if `raise_err` is True and any relative path is not part of the directory tree.
+
+        """
+        if not paths:
+            return root
+        # Make sure paths is a list
+        if isinstance(paths, str):
+            paths = [paths]
+        pruned_tree = {}  # type: Dict
+        for rel_path in paths:
+            # Normalize relative path to ensure reduce() finds it correctly
+            dirnames = os.path.normpath(rel_path).split(os.path.sep)
+            try:
+                tree_node = reduce(operator.getitem, dirnames, root)
+            except KeyError:
+                if raise_err:
+                    # Suppress exception context to display only the ValueError
+                    raise ValueError("Path '{}' not found in the directory tree".format(rel_path)) from None
+                continue
+            # Create/find the path in the pruned tree and attach the subtree
+            pruned_node = pruned_tree
+            for name in dirnames[:-1]:
+                pruned_node = pruned_node.setdefault(name, {})
+            pruned_node[dirnames[-1]] = tree_node
+        return pruned_tree
+
+    @staticmethod
+    def _eval_tree(root: Dict, test_func: Callable) -> Iterator[str]:
+        """Yields each file in the directory tree for which the test function returns True.
+
+        Args:
+            test_func: Test function to apply to each file. It has to match the following interface::
+
+                def test_func(file: str) -> bool:
+                    ...
+
+        """
+        nodes_left = [('', root)]
```

Comment:
Unless you think it improves the readability of the code making `nodes_left` a FIFO structure, `list.pop()` has "constant" cost, and a `deque` requires slightly more memory than a list.

Reply:
It's not about readability, it's about performance / purpose.
You're right that `list.pop()` and `list.append()` have a constant time on average, but the underlying C array still needs to grow / shrink sometimes which would involve a `realloc`. Similarly, the deque object may have to allocate/deallocate new list elements, but not the entire list, making the `pop()` and `append()` operations much more predictable in terms of runtime. I've run some tests, and the performance of `deque` is very slightly faster than `list`. Not enough to matter given the size of citest, though.

Reply:
I see. That's interesting. And here comes the question that sometimes ask myself: we are measuring this for CITest, but we always talk about how other teams may use this. Is it worth doing the change just in case they handle larger datasets?

Reply:
Fair point. We may ourselves use this module to replace the script that compares the FTP directories, which has an ever-growing number of files, or do other things on large directory structures.
In the end, it's only one change (the constructor), since the interface (`pop()` / `append()`) is the same